server/src/main/java/org/apache/lucene/index/BaseSingleStarTreeBuilder.java

```diff
@@ -0,0 +1,660 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.apache.lucene.index;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.codecs.DocValuesConsumer;
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.opensearch.index.compositeindex.datacube.DateDimension;
+import org.opensearch.index.compositeindex.datacube.Dimension;
+import org.opensearch.index.compositeindex.datacube.Metric;
+import org.opensearch.index.compositeindex.datacube.MetricStat;
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeField;
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeFieldConfiguration;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.MetricStatFieldPair;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.ValueAggregator;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.ValueAggregatorFactory;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.numerictype.StarTreeNumericType;
+import org.opensearch.index.compositeindex.datacube.startree.builder.SingleTreeBuilder;
+import org.opensearch.index.compositeindex.datacube.startree.builder.StarTreeDocValuesIteratorFactory;
+import org.opensearch.index.compositeindex.datacube.startree.data.StarTreeDocument;
+import org.opensearch.index.compositeindex.datacube.startree.utils.StarTreeBuilderUtils;
+import org.opensearch.index.fielddata.IndexNumericFieldData;
+import org.opensearch.index.mapper.Mapper;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.index.mapper.NumberFieldMapper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Base class for star-tree builder
+ */
+public abstract class BaseSingleStarTreeBuilder implements SingleTreeBuilder {
+
+    // TODO: STAR_TREE_CODEC will be moved to CodecService once the Star Tree Codec is defined
+    public static final String STAR_TREE_CODEC = "startreecodec";
+
+    private static final Logger logger = LogManager.getLogger(BaseSingleStarTreeBuilder.class);
+
+    public static final int STAR_IN_DOC_VALUES_INDEX = -1;
```

Comment:
What if user gives '-1' as part of their documents ? As per our discussions, maybe min value, max value helps a bit better, but it doesn't work well in case of delta encoding etc.

Reply:
Agreed, but not sure if min / max value are also full proof though. We can see if Double.NaN is permissible by Lucene.

Reply:
As discussed, will be adding an additional flag to star tree node to denote that the star node